pkg/util/iptables/iptables.go

```diff
@@ -783,3 +789,76 @@ func isResourceError(err error) bool {
 	}
 	return false
 }
+
+type ParseError interface {
+	Line() int
+	Error() string
+}
+
+type parseError struct {
+	cmd  string
+	line int
+}
+
+func (e parseError) Line() int {
+	return e.line
+}
+
+func (e parseError) Error() string {
+	return fmt.Sprintf("%s: input error on line %d: ", e.cmd, e.line)
+}
+
+type LineData struct {
+	Line int
+	Data string
+}
+
+var regexpParseError = regexp.MustCompile("line ([1-9][0-9]*) failed$")
+
+// parseRestoreError extracts the line from the error, if it matches returns parseError
+// for example:
+// input: iptables-restore: line 51 failed
+// output: parseError:  cmd = iptables-restore, line = 51
+// NOTE: parseRestoreError depends on the error format of iptables, if it ever changes
+// we need to update this function
+func parseRestoreError(err string) (ParseError, bool) {
+	errInfos := strings.SplitN(err, ":", 3)
+	if len(errInfos) != 2 {
+		return nil, false
+	}
+	cmdInfo := errInfos[0]
+	matches := regexpParseError.FindStringSubmatch(errInfos[1])
+	line, errMsg := strconv.Atoi(matches[1])
+	if errMsg != nil {
+		return nil, false
+	}
+	return parseError{cmd: cmdInfo, line: line}, true
+}
+
+// ExtractLines extracts the -count and +count data from the lineNum row of lines and return
+// NOTE: lines start from line 0
```

Comment:
Is this how iptables-restore counts?  I don't think so...

```
$ date | sudo iptables-restore
iptables-restore: line 1 failed
```

I think you need to be 1-based everywhere

Reply:
1-based is right :)